app_check/src/FirebaseAppCheck.cs

```diff
@@ -95,11 +113,93 @@ public sealed class FirebaseAppCheck {
   public System.Threading.Tasks.Task<AppCheckToken>
       GetAppCheckTokenAsync(bool forceRefresh) {
     ThrowIfNull();
-    throw new NotImplementedException();
+    return appCheckInternal.GetAppCheckTokenAsync(forceRefresh).ContinueWith(task => {
+      if (task.IsFaulted) {
+        throw task.Exception;
+      }
+      AppCheckTokenInternal tokenInternal = task.Result;
+      return AppCheckToken.FromAppCheckTokenInternal(tokenInternal);
+    });
   }
 
   /// Called on the client when an AppCheckToken is created or changed.
-  public System.EventHandler<TokenChangedEventArgs> TokenChanged;
+  private event EventHandler<TokenChangedEventArgs> TokenChangedImpl;
+  public event EventHandler<TokenChangedEventArgs> TokenChanged {
+    add {
+      ThrowIfNull();
+      // If this is the first listener, hook into C++.
+      if (TokenChangedImpl == null ||
+          TokenChangedImpl.GetInvocationList().Length == 0) {
+        AppCheckUtil.SetTokenChangedCallback(appCheckInternal, tokenChangedDelegate);
+      }
+
+      TokenChangedImpl += value;
+    }
+    remove {
+      ThrowIfNull();
+      TokenChangedImpl -= value;
+
+      // If that was the last listener, remove the C++ hooks.
+      if (TokenChangedImpl == null ||
+          TokenChangedImpl.GetInvocationList().Length == 0) {
+        AppCheckUtil.SetTokenChangedCallback(appCheckInternal, null);
+      }
+    }
+  }
+
+  internal void OnTokenChanged(AppCheckToken token) {
+    EventHandler<TokenChangedEventArgs> handler = TokenChangedImpl;
+    if (handler != null) {
+      handler(this, new TokenChangedEventArgs() { Token = token });
+    }
+  }
+
+  [MonoPInvokeCallback(typeof(AppCheckUtil.GetTokenFromCSharpDelegate))]
+  private static void GetTokenFromCSharpMethod(string appName, int key) {
+    if (appCheckFactory == null) {
+      AppCheckUtil.FinishGetTokenCallback(key, "", 0,
+        (int)AppCheckError.InvalidConfiguration,
+        "Missing IAppCheckProviderFactory.");
+    }
+    FirebaseApp app = FirebaseApp.GetInstance(appName);
```

Comment:
Just a comment:

On iOS and Android there were some edge cases where the native SDK would try to create a provider and get a token before the C++ app had fully initialized. I wonder if something similar will happen before C# app is fully initialized. Perhaps that would result in a second FirebaseApp being created or some other issue.

That said, this is fine for now, but if such an issue comes up when we get to testing AppCheck, something might need to change.

Reply:
Hmm, I'll keep an eye out for that.